e2e/thin_provisioning_test.go

```diff
@@ -0,0 +1,274 @@
+				volumeName, err = getVolumeNameofPVC("thinvol"+num, ns)
+				return err
+			}).Should(Succeed())
+
+			var lv *thinlvinfo
+			Eventually(func() error {
+				fmt.Println("Testing", "getThinLVInfo", "1")
+				lv, err = getThinLVInfo(volumeName)
+				return err
+			}).Should(Succeed())
+
+			vgName := "node2-myvg4"
+			if isDaemonsetLvmdEnvSet() {
+				vgName = "node-myvg5"
+			}
+			Expect(vgName).Should(Equal(lv.vgName))
+
+			poolName := "pool0"
+			Expect(poolName).Should(Equal(lv.poolName))
+		}
+	})
+
+	It("should check overprovision limits", func() {
+		By("Deploying a PVC to use up the available thinpoolsize * overprovisioning")
+
+		nodeName := "topolvm-e2e-worker3"
+		if isDaemonsetLvmdEnvSet() {
+			nodeName = getDaemonsetLvmdNodeName()
+		}
+
+		thinPvcYAML := []byte(fmt.Sprintf(thinPVCTemplateYAML, "thinvol", "18"))
+		stdout, stderr, err := kubectlWithInput(thinPvcYAML, "apply", "-n", ns, "-f", "-")
+		Expect(err).ShouldNot(HaveOccurred(), "stdout=%s, stderr=%s", stdout, stderr)
+
+		thinPodYAML := []byte(fmt.Sprintf(thinPodTemplateYAML, "thinpod", "thinvol", nodeName))
+		stdout, stderr, err = kubectlWithInput(thinPodYAML, "apply", "-n", ns, "-f", "-")
+		Expect(err).ShouldNot(HaveOccurred(), "stdout=%s, stderr=%s", stdout, stderr)
+
+		var volumeName string
+		Eventually(func() error {
+			volumeName, err = getVolumeNameofPVC("thinvol", ns)
+			return err
+		}).Should(Succeed())
+
+		var lv *thinlvinfo
+		Eventually(func() error {
+			fmt.Println("Testing", "getThinLVInfo", "1")
+			lv, err = getThinLVInfo(volumeName)
+			return err
+		}).Should(Succeed())
+
+		vgName := "node3-myvg4"
+		if isDaemonsetLvmdEnvSet() {
+			vgName = "node-myvg5"
+		}
+		Expect(vgName).Should(Equal(lv.vgName))
+
+		poolName := "pool0"
+		Expect(poolName).Should(Equal(lv.poolName))
+
+		By("Failing to deploying a PVC when total size > thinpoolsize * overprovisioning")
+		thinPvcYAML = []byte(fmt.Sprintf(thinPVCTemplateYAML, "thinvol2", "5"))
+		stdout, stderr, err = kubectlWithInput(thinPvcYAML, "apply", "-n", ns, "-f", "-")
+		Expect(err).ShouldNot(HaveOccurred(), "stdout=%s, stderr=%s", stdout, stderr)
+
+		thinPodYAML = []byte(fmt.Sprintf(thinPodTemplateYAML, "thinpod2", "thinvol2", nodeName))
+		stdout, stderr, err = kubectlWithInput(thinPodYAML, "apply", "-n", ns, "-f", "-")
+		Expect(err).ShouldNot(HaveOccurred(), "stdout=%s, stderr=%s", stdout, stderr)
+
+		Eventually(func() error {
+			stdout, stderr, err := kubectl("get", "-n", ns, "pvc", "thinvol2", "-o", "json")
+			if err != nil {
+				return fmt.Errorf("failed to get PVC. stdout: %s, stderr: %s, err: %v", stdout, stderr, err)
+			}
+
+			var pvc corev1.PersistentVolumeClaim
+			err = json.Unmarshal(stdout, &pvc)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshal PVC. stdout: %s, err: %v", stdout, err)
+			}
+			if pvc.Status.Phase == corev1.ClaimBound {
+				return fmt.Errorf("PVC should not be bound")
+			}
+			/*
+				volName, err := getVolumeNameofPVC("thinvol2", ns)
```

Comment:
I could not get this to work because the LogicalVolume gets deleted when there is insufficient space. Is that the expected behaviour?

Reply:
> Is that the expected behaviour?

Yes, it is. When a volume failed to be created, the corresponding LogicalVolume CR is deleted in the following path.
[CreateVolume()](https://github.com/topolvm/topolvm/blob/main/driver/controller.go#L197) of controllerService
-> [CreateVolume()](https://github.com/topolvm/topolvm/blob/main/driver/k8s/logicalvolume_service.go#L176) of LogicalVolumeService
-> [waitForStatusUpdate()](https://github.com/topolvm/topolvm/blob/main/driver/k8s/logicalvolume_service.go#L406) of LogicalVolumeService

When volume creation fails, `volume_id` is not returned to the CSI sidecar. Because `volume_id` is required to delte volumes, the LogicalVolume would be leaked. To prevent it, the LogicalVolume is deleted on the failure of LogicalVolume creation.

Reply:
Understood. Thank you for the explanation.